### PR TITLE
Switching to the new MARC API

### PR DESCRIPTION
**Pull request by shadeMe**

### CODE REVIEW

File: cpp/add_isbns_or_issns_to_articles.cc [hunk 1]
```diff
@@ -92,16 +91,19 @@ void PopulateParentIdToISBNAndISSNMap(
...
             }
         }
     }
 
     if (not err_msg.empty())
-        logger->error(err_msg);
+        LOG_ERROR(err_msg);
 
     if (verbose) {
         std::cerr << "Read " << count << " records.\n";
```
Comment (ruschein) [inline cpp/add_isbns_or_issns_to_articles.cc]: Please use LOG_INFO here.

File: cpp/add_isbns_or_issns_to_articles.cc [hunk 2]
```diff
@@ -111,44 +113,42 @@ void PopulateParentIdToISBNAndISSNMap(
...
 
 
-void AddMissingISBNsOrISSNsToArticleEntries(const bool verbose, MarcReader * const marc_reader,
-                                            MarcWriter * const marc_writer, const std::unordered_map<std::string,
+void AddMissingISBNsOrISSNsToArticleEntries(const bool verbose, MARC::Reader * const marc_reader,
+                                            MARC::Writer * const marc_writer, const std::unordered_map<std::string,
                                             std::string> &parent_id_to_isbn_and_issn_map)
 {
     if (verbose)
         std::cout << "Starting augmentation of article entries.\n";
```
Comment (ruschein) [inline cpp/add_isbns_or_issns_to_articles.cc]: LOG_INFO

File: cpp/add_referenceterms.cc [hunk 3]
```diff
@@ -82,31 +79,32 @@ void ProcessRecord(MarcRecord * const record, const std::string &output_tag_and_
...
     const std::string tag(GetTag(output_tag_and_subfield_code));
-    if (record->getFieldIndex(tag) != MarcRecord::FIELD_NOT_FOUND)
-        logger->error("Field with tag " + tag + " is not empty for PPN " + record->getControlNumber() + '\n');
+    if (record->hasFieldWithTag(tag))
+        LOG_ERROR("Field with tag " + tag + " is not empty for PPN " + record->getControlNumber() + '\n');
     std::string subfield_spec(GetSubfieldCodes(output_tag_and_subfield_code));
     if (subfield_spec.size() != 1)
-        logger->error("We currently only support a single subfield and thus specifying " + subfield_spec
+        LOG_ERROR("We currently only support a single subfield and thus specifying " + subfield_spec
                       + " as output subfield is not valid\n");
```
Comment (ruschein) [inline cpp/add_referenceterms.cc]: Indentation.

File: cpp/add_referenceterms.cc [hunk 4]
```diff
@@ -82,31 +79,32 @@ void ProcessRecord(MarcRecord * const record, const std::string &output_tag_and_
...
-        logger->warning("Could not insert field " + tag + " for PPN " + record->getControlNumber());
-    ++modified_count;
+    if (not record->insertField(tag, subfields))
+        LOG_WARNING("Could not insert field " + tag + " for PPN " + record->getControlNumber());
+    ++*modified_count;
 }
 
 
-void InsertSynonyms(MarcReader * const marc_reader, MarcWriter * marc_writer,
+void InsertSynonyms(MARC::Reader * const marc_reader, MARC::Writer * marc_writer,
```
Comment (ruschein) [inline cpp/add_referenceterms.cc]: marc_writer should be a const pointer.

File: cpp/add_referenceterms.cc [hunk 5]
```diff
@@ -141,6 +140,8 @@ int main(int argc, char **argv) {
         ExtractSynonyms(reference_data_id_term_list_input.get(), &synonym_map);
         InsertSynonyms(marc_reader.get(), marc_writer.get(), TITLE_DATA_UNUSED_FIELD_FOR_SYNONYMS, synonym_map);
     } catch (const std::exception &x) {
```
Comment (ruschein) [inline cpp/add_referenceterms.cc]: No try-catch, please.

File: cpp/add_title_keywords.cc [hunk 6]
```diff
@@ -82,34 +82,26 @@ void FilterOutStopwords(const std::unordered_set<std::string> &stopwords,
...
 
 
-void AugmentKeywordsWithTitleWords(const bool verbose, MarcReader * const marc_reader,
-    MarcWriter * const marc_writer,
+void AugmentKeywordsWithTitleWords(const bool verbose, MARC::Reader * const marc_reader,
+    MARC::Writer * const marc_writer,
     const std::map<std::string, std::unordered_set<std::string>> &language_codes_to_stopword_sets)
 {
     if (verbose)
         std::cerr << "Starting augmentation of stopwords.\n";
```
Comment (ruschein) [inline cpp/add_title_keywords.cc]: LOG_INFO

File: cpp/add_isbns_or_issns_to_articles.cc [hunk 7]
```diff
@@ -196,19 +198,17 @@ int main(int argc, char **argv) {
...
+
+    auto marc_reader(MARC::Reader::Factory(marc_input_filename));
+    auto marc_writer(MARC::Writer::Factory(marc_output_filename));
+
+    std::unordered_map<std::string, std::string> parent_id_to_isbn_and_issn_map;
+    PopulateParentIdToISBNAndISSNMap(verbose, marc_reader.get(), &parent_id_to_isbn_and_issn_map);
+    marc_reader->rewind();
+
+    AddMissingISBNsOrISSNsToArticleEntries(verbose, marc_reader.get(), marc_writer.get(),
+                                            parent_id_to_isbn_and_issn_map);
```
Comment (ruschein) [inline cpp/add_isbns_or_issns_to_articles.cc]: Indentation.

File: cpp/add_title_keywords.cc [hunk 8]
```diff
@@ -182,30 +173,31 @@ int main(int argc, char **argv) {
     const std::string marc_input_filename(argv[verbose ? 2 : 1]);
     const std::string marc_output_filename(argv[verbose ? 3 : 2]);
     if (unlikely(marc_input_filename == marc_output_filename))
-        logger->error("MARC input file name equals MARC output file name!");
+        LOG_ERROR("MARC input file name equals MARC output file name!");
 
-    std::unique_ptr<MarcReader> marc_reader(MarcReader::Factory(marc_input_filename, MarcReader::BINARY));
-    std::unique_ptr<MarcWriter> marc_writer(MarcWriter::Factory(marc_output_filename, MarcWriter::BINARY));
+    auto marc_reader(MARC::Reader::Factory(marc_input_filename));
```
Comment (ruschein) [inline cpp/add_title_keywords.cc]: Do we know why BINARY was hardcoded here?  Maybe it's part of a pipeline reading from a FIFO?

File: cpp/augment_bible_references.cc [hunk 9]
```diff
@@ -68,28 +63,26 @@ void LoadBibleOrderMap(const bool verbose, File * const input,
...
 
 
 /* Pericopes are found in 130$a if there are also bible references in the 430 field. You should therefore
    only call this after acertaining that one or more 430 fields contain a bible reference. */
-bool FindPericopes(const MarcRecord &record, const std::set<std::pair<std::string, std::string>> &ranges,
+bool FindPericopes(const MARC::Record &record, const std::set<std::pair<std::string, std::string>> &ranges,
                    std::unordered_multimap<std::string, std::string> * const pericopes_to_ranges_map)
 {
-    static const std::string PERICOPE_FIELD("130");
+    // ### 130 doesn't seem like it's repeatable, why does the code below think otherwise?
```
Comment (ruschein) [inline cpp/augment_bible_references.cc]: This comment seems no longer relevant.

File: cpp/augment_bible_references.cc [hunk 10]
```diff
@@ -232,25 +225,23 @@ static unsigned unknown_book_count;
...
+bool GetBibleRanges(const std::string &field_tag, const MARC::Record &record,
                     const std::unordered_set<std::string> &books_of_the_bible,
                     const std::unordered_map<std::string, std::string> &bible_book_to_code_map,
                     std::set<std::pair<std::string, std::string>> * const ranges)
 {
     ranges->clear();
 
-    size_t index(record.getFieldIndex(field_tag));
-    if (index == MarcRecord::FIELD_NOT_FOUND)
+    const auto field_range(record.getTagRange(field_tag));
```
Comment (ruschein) [inline cpp/augment_bible_references.cc]: I don't think we need an explicit early check for an empty range.

File: cpp/augment_bible_references.cc [hunk 11]
```diff
@@ -396,28 +382,26 @@ bool FindGndCodes(const bool verbose, const std::string &tags, const MarcRecord
 
     bool found_at_least_one(false);
     for (const auto &tag : individual_tags) {
-        for (size_t index(record.getFieldIndex(tag));
-             index < record.getNumberOfFields() and record.getTag(index) == tag; ++index)
-        {
-            const Subfields subfields(record.getSubfields(index));
-            const std::string subfield2(subfields.getFirstSubfieldValue('2'));
+        const auto field_range(record.getTagRange(tag));
```
Comment (ruschein) [inline cpp/augment_bible_references.cc]: If you con't use field_range for anything but the loop below, please eliminate it!

File: cpp/augment_bible_references.cc [hunk 12]
```diff
@@ -457,62 +440,51 @@ void AugmentBibleRefs(const bool verbose, MarcReader * const marc_reader, MarcWr
                 }
 
                 // Put the data into the $a subfield:
-                record.insertSubfield(BibleUtil::BIB_REF_RANGE_TAG, 'a', range_string);
+                bible_reference_tag_field->insertOrReplaceSubfield('a', range_string);
```
Comment (ruschein) [inline cpp/augment_bible_references.cc]: We have established that bible_reference_tag_field is record.end()) on line 429!  You need to use Record::insertField().

File: cpp/generate_title_and_author_list.cc [hunk 13]
```diff
@@ -23,50 +23,42 @@
...
-        const std::string _245_contents(record.getFieldData("245"));
-        if (unlikely(_245_contents.empty()))
+        const auto field_245(record.findTag("245"));
+        if (unlikely(field_245 == record.end()))
             continue;
 
-        const Subfields subfields(_245_contents);
-        const std::string main_title(subfields.getFirstSubfieldValue('a'));
+        const auto subfields(field_245->getSubfields());
+        const std::string main_title(subfields.getFirstSubfieldWithCode('a'));
```
Comment (ruschein) [inline cpp/generate_title_and_author_list.cc]: I don't think we need "subfields".  I think you can use field_245->getFirstSubfieldWithCode('a') directly.

File: cpp/marc_compare.cc [hunk 14]
```diff
@@ -20,63 +20,73 @@
...
+        if (lhs.getNumberOfFields() != rhs.getNumberOfFields()) {
+            LOG_ERROR("Number of fields (" + lhs.getControlNumber() + "):\nLHS: "
+                      + std::to_string(lhs.getNumberOfFields()) + "\nRHS: " + std::to_string(rhs.getNumberOfFields()));
+        }
 
-        for (size_t index(0); index < lhs.getNumberOfFields(); ++index) {
-            if (lhs.getTag(index) != rhs.getTag(index))
-                logger->error("Tag mismatch (" + lhs.getControlNumber() + "):\nLHS: " + lhs.getTag(index).to_string()
-                      + "\nRHS: " + rhs.getTag(index).to_string());
+        for (auto lhs_field(lhs.begin()), rhs_field(rhs.begin());
```
Comment (ruschein) [inline cpp/marc_compare.cc]: I don't think we need 3 lines for the loop header.

File: cpp/add_title_keywords.cc (hunk 8 above)
Comment (shadeMe) [reply cpp/add_title_keywords.cc]: Apparently, this tool is never used.

File: cpp/add_isbns_or_issns_to_articles.cc [hunk 15]
```diff
@@ -47,43 +46,43 @@ void Usage() {
 
 
 void PopulateParentIdToISBNAndISSNMap(
-    const bool verbose, MarcReader * const marc_reader,
+    const bool verbose, MARC::Reader * const marc_reader,
     std::unordered_map<std::string, std::string> * const parent_id_to_isbn_and_issn_map)
 {
     if (verbose)
         std::cout << "Starting extraction of ISBN's and ISSN's.\n";
```
Comment (ruschein) [inline cpp/add_isbns_or_issns_to_articles.cc]: LOG_INFO

File: cpp/add_isbns_or_issns_to_articles.cc (hunk 1 above)
Comment (ruschein) [inline cpp/add_isbns_or_issns_to_articles.cc]: LOG_INFO

File: cpp/add_isbns_or_issns_to_articles.cc [hunk 16]
```diff
@@ -196,19 +198,21 @@ int main(int argc, char **argv) {
...
-        logger->error("Master input file name equals output file name!");
+        LOG_ERROR("Master input file name equals output file name!");
 
-    std::unique_ptr<MarcReader> marc_reader(MarcReader::Factory(marc_input_filename, MarcReader::BINARY));
-    std::unique_ptr<MarcWriter> marc_writer(MarcWriter::Factory(marc_output_filename, MarcWriter::BINARY));
+    auto marc_reader(MARC::Reader::Factory(marc_input_filename));
+    auto marc_writer(MARC::Writer::Factory(marc_output_filename));
 
     try {
         std::unordered_map<std::string, std::string> parent_id_to_isbn_and_issn_map;
```
Comment (ruschein) [inline cpp/add_isbns_or_issns_to_articles.cc]: Please take out the try and the catch clause.

**Pull request merged**